In [167]:
import pypsa
import pandas as pd

# Process and analysis

## Network without NEP

In [317]:
n_no_nep = pypsa.Network("../results/test_scenario_wo_tyndp_nep/prenetworks-brownfield/base_s_50_lv1.5___2035.nc")

INFO:pypsa.io:Imported network base_s_50_lv1.5___2035.nc has buses, carriers, generators, global_constraints, lines, links, loads, storage_units, stores


In [318]:
ac_lines = n_no_nep.lines

In [319]:
line_cap = ac_lines.s_nom.mul(ac_lines.length).div(1e3).sum()
line_cap

128789.57305278498

In [324]:
dc_links = n_no_nep.links.query("carrier=='DC'")
dc_links.head()

bus0    bus1 type carrier  efficiency  active  \
Link                                                                  
relation/8185420+2   NL0 0   DK0 0           DC    0.966905    True   
relation/2505320+2   DK1 0  DE0 22           DC    0.973812    True   
relation/5487095+2   DK0 0   DK1 0           DC    0.975075    True   
relation/8193755+2  DE0 15   BE0 0           DC    0.976273    True   
relation/3392010+2   SE1 0   PL0 0           DC    0.950468    True   

                    build_year  lifetime   p_nom  p_nom_mod  ...  efficiency2  \
Link                                                         ...                
relation/8185420+2           0       inf   700.0        0.0  ...          1.0   
relation/2505320+2           0       inf   600.0        0.0  ...          1.0   
relation/5487095+2           0       inf   600.0        0.0  ...          1.0   
relation/8193755+2           0       inf  1000.0        0.0  ...          1.0   
relation/3392010+2           0       inf   600.0        0.0  ...          1.0   

                    under_construction  voltage  underwater_fraction  \
Link                                                                   
relation/8185420+2                 0.0      NaN             0.926393   
relation/2505320+2                 0.0      NaN             0.281098   
relation/5487095+2                 0.0      NaN             0.563251   
relation/8193755+2                 0.0      NaN             0.000000   
relation/3392010+2                 0.0      NaN             0.941267   

                    reversed  geometry  underground  length_original  dc  \
Link                                                                       
relation/8185420+2     False       NaN          0.0       578.122730 NaN   
relation/2505320+2     False       NaN          0.0       272.242067 NaN   
relation/5487095+2     False       NaN          0.0       216.510855 NaN   
relation/8193755+2     False       NaN          0.0       163.740720 NaN   
relation/3392010+2     False       NaN          0.0      1314.985161 NaN   

                    location  
Link                          
relation/8185420+2            
relation/2505320+2            
relation/5487095+2            
relation/8193755+2            
relation/3392010+2            

[5 rows x 49 columns]

In [325]:
dc_links_cap = dc_links.p_nom.mul(dc_links.length).div(1e3).sum() # TWkm
dc_links_cap

2880.2877529052575

In [326]:
grid_cap_wo_nep = line_cap + dc_links_cap
grid_cap_wo_nep

131669.86080569023

## Network with NEP

In [358]:
n_nep = pypsa.Network("../results/test_scenario_w_tyndp_nep/prenetworks-brownfield/base_s_50_lv1.5___2035.nc")

INFO:pypsa.io:Imported network base_s_50_lv1.5___2035.nc has buses, carriers, generators, global_constraints, lines, links, loads, storage_units, stores


In [359]:
n_nep.statistics.installed_capacity()

component    carrier                                  
StorageUnit  Pumped Hydro Storage                           30373.37000
             Reservoir & Dam                                34862.00720
Line         AC                                            683254.62583
Link         Combined-Cycle Gas                             99480.59829
             DC                                             97560.00000
             Open-Cycle Gas                                    89.90361
             coal                                           27700.77775
             lignite                                        67353.58717
             nuclear                                        72996.93252
             oil                                             2658.70986
             oil refining                                 1000000.00000
             residential rural gas boiler                   60327.55165
             residential rural ground heat pump              3213.25104
         

In [360]:
ac_lines = n_nep.lines

In [361]:
line_cap = ac_lines.s_nom.mul(ac_lines.length).div(1e3).sum()
line_cap

143041.59037987786

In [362]:
dc_links = n_nep.links.query("carrier=='DC'")
dc_links.head()

bus0    bus1 type carrier  efficiency  active  build_year  \
Link                                                                        
TYNDP2020_1    DE0 3  DE0 25           DC    0.962911    True        2026   
TYNDP2020_10  DE0 21  DE0 16           DC    0.972774    True        2024   
TYNDP2020_22  DE0 19  DE0 14           DC    0.969353    True        2025   
TYNDP2020_23  DE0 17  DE0 21           DC    0.973618    True        2025   
DC21          DE0 17  DE0 24           DC    0.976574    True        2032   

              lifetime   p_nom  p_nom_mod  ...  voltage  underwater_fraction  \
Link                                       ...                                 
TYNDP2020_1        inf  2000.0        0.0  ...      NaN                 0.00   
TYNDP2020_10       inf  2000.0        0.0  ...      NaN                 0.00   
TYNDP2020_22       inf  2000.0        0.0  ...      NaN                 0.00   
TYNDP2020_23       inf  2000.0        0.0  ...      NaN                 0.04   
DC21               inf  2000.0        0.0  ...      NaN                 0.03   

                                                           tags  reversed  \
Link                                                                        
TYNDP2020_1   "url"=>"https://tyndp2020-project-platform.azu...     False   
TYNDP2020_10  "url"=>"https://tyndp2020-project-platform.azu...     False   
TYNDP2020_22  "url"=>"https://tyndp2020-project-platform.azu...     False   
TYNDP2020_23  "url"=>"https://tyndp2020-project-platform.azu...     False   
DC21          {"url":"https://data.netzausbau.de/2037-2023/N...     False   

                                                       geometry  underground  \
Link                                                                           
TYNDP2020_1   LINESTRING (9.24087500000001 53.906765, 9.1221...          1.0   
TYNDP2020_10  LINESTRING (6.58493 51.245584, 8.478699 49.283...          0.0   
TYNDP2020_22  LINESTRING (11.637268 52.27404, 12.292328 48.6...          1.0   
TYNDP2020_23  LINESTRING (7.23587000000001 53.374317, 6.5849...          0.0   
DC21          LINESTRING (8.106301 53.5278793, 8.0787633 51....          1.0   

              length_original  dc  project_status  location  
Link                                                         
TYNDP2020_1        756.021733 NaN   in_permitting            
TYNDP2020_10       318.066200 NaN   in_permitting            
TYNDP2020_22       469.463432 NaN   in_permitting            
TYNDP2020_23       280.773609 NaN   in_permitting            
DC21               150.516725 NaN       confirmed            

[5 rows x 51 columns]

In [363]:
dc_links_cap = dc_links.p_nom.mul(dc_links.length).div(1e3).sum() # TWkm
dc_links_cap

24148.836153693133

In [364]:
grid_cap_w_nep = line_cap + dc_links_cap
grid_cap_w_nep

167190.42653357098

In [365]:
# Ratio
grid_cap_w_nep/grid_cap_wo_nep

1.2697699041415382

## Carbon emission targets

In [369]:
ctrys = ['AT', 'BE', 'CH', 'CZ', 'DE', 'DK', 'FR', 'IT', 'LU', 'NL', 'PL', 'SE']

In [370]:
# 1990 emissions in MtCO2
co2_totals = pd.read_csv("../resources/co2_totals_1990.csv", index_col=0).loc[ctrys]

In [373]:
# 2005 emissions in MtCO2
co2_totals_2005 = pd.read_csv("../resources/co2_totals_2005.csv", index_col=0).loc[ctrys]

In [374]:
# Total emissions excl. LULUCF
co2_totals_excl_lulucf = co2_totals.drop(["LULUCF"], axis=1).sum(axis=1)
co2_totals_excl_lulucf

Country_code
AT      62.124701
BE     120.309343
CH      44.534441
CZ     166.068823
DE    1052.348544
DK      54.685986
FR     400.777042
IT     438.008802
LU      11.847635
NL     163.302301
PL     376.546518
SE      57.348825
dtype: float64

In [375]:
# Total emissions incl. LULUCF
co2_totals_incl_lulucf = co2_totals.sum(axis=1)
co2_totals_incl_lulucf

Country_code
AT      49.968104
BE     117.060288
CH      42.515838
CZ     160.298010
DE    1021.884333
DK      60.858777
FR     374.572263
IT     432.346810
LU      11.927560
NL     169.786966
PL     343.554029
SE      21.112932
dtype: float64

In [376]:
# Total emissions excl. LULUCF
co2_totals_excl_lulucf_2005 = co2_totals_2005.drop(["LULUCF"], axis=1).sum(axis=1)
co2_totals_excl_lulucf_2005

Country_code
AT     79.192432
BE    125.661688
CH     45.915746
CZ    126.738399
DE    866.388549
DK     52.164863
FR    426.412463
IT    500.005676
LU     12.105448
NL    177.897237
PL    323.161267
SE     53.852857
dtype: float64

In [377]:
# Total emissions incl. LULUCF 2005
co2_totals_incl_lulucf_2005 = co2_totals_2005.sum(axis=1)
co2_totals_incl_lulucf_2005

Country_code
AT     68.414161
BE    124.024742
CH     43.712103
CZ    119.316791
DE    851.168952
DK     57.805099
FR    377.355895
IT    463.962937
LU     11.462305
NL    183.428661
PL    271.748435
SE     19.936933
dtype: float64

### ESR targets 2030 (compared to 2005) [Source](https://climate.ec.europa.eu/eu-action/effort-sharing-member-states-emission-targets/effort-sharing-2021-2030-targets-and-flexibilities_en)

In [378]:
esr_targets = {
    "AT": 0.48,
    "BE": 0.47,
    "CZ": 0.26,
    "DE": 0.5,
    "DK": 0.5,
    "FR": 0.475,
    "IT": 0.437,
    "LU": 0.5,
    "NL": 0.48,
    "PL": 0.177,
    "SE": 0.5,
}

### Country specific emission reduction targets

| Country | 2030 | 2035 | 2040 | 2045 | 2050 | Source |
|---------|------|------|------|------|------|--------|
| DE | 65% below 1990 excl. LULUCF | - | 85% below 1990 excl. LULUCF | 100% | - | [Source](https://www.bundesregierung.de/breg-de/themen/klimaschutz/klimaschutzgesetz-2021-1913672) |
| AT | 48% below 2005 excl. LULUCF | - | 100% | - | - | [Source1](https://climate.ec.europa.eu/eu-action/effort-sharing-member-states-emission-targets/effort-sharing-2021-2030-targets-and-flexibilities_en) [Source2](https://www.sustainability.gov/pdfs/austria-nzgi-roadmap.pdf) |
| BE | 55% below 1990 excl. LULUCF | - | - | - | 100% | [Source](https://www.sustainability.gov/pdfs/belgium-nzgi-roadmap.pdf) |
| CH | 50% below 1990 excl. LULUCF | - | - |  - | 100% | [Source](https://www.bafu.admin.ch/bafu/en/home/topics/climate/info-specialists/emission-reduction/reduction-targets.html) |
| CZ | 26% below 2005 excl. LULUCF | - | - |  - | 100% | [Source](https://tracker.carbongap.org/region/czech-republic/) |
| DK | 70% below 1990 incl. LULUCF and 50% below 2005 excl. LULUCF | - | - |  100% | 110% | [Source](https://unfccc.int/sites/default/files/resource/SBI60_MA_DNK-7June2024-Presentation.pdf) |
| FR | 47.5% below 2005 excl. LULUCF | - | - |  - | 100% | [Source](https://climate.ec.europa.eu/eu-action/effort-sharing-member-states-emission-targets/effort-sharing-2021-2030-targets-and-flexibilities_en) |
| IT | 43.7% below 2005 excl. LULUCF | - | - |  - | 100% | [Source1](https://climate.ec.europa.eu/eu-action/effort-sharing-member-states-emission-targets/effort-sharing-2021-2030-targets-and-flexibilities_en) [Source2](https://www.iea.org/countries/italy#) |
| LU | 55% below 2005 excl. LULUCF | - | - |  - | 100% | [Source](https://gouvernement.lu/en/dossiers/2023/2023-pnec.html) |
| NL | 55% below 1990 | - | - |  - | 100% | [Source](https://commission.europa.eu/document/download/b6d21e56-4297-4b91-a692-300716209f72_en?filename=NL_FINAL%20UPDATED%20NECP%202021-2030%20%28English%29.pdf) |
| PL | 17.7% below 2005 excl. LULUCF | - | - |  - |  | [Source](https://climate.ec.europa.eu/eu-action/effort-sharing-member-states-emission-targets/effort-sharing-2021-2030-targets-and-flexibilities_en) |
| SE | 63% below 1990 | - | 75% below 1990 |  100% |  | [Source](https://www.bafu.admin.ch/bafu/en/home/topics/climate/info-specialists/emission-reduction/reduction-targets.html) |

In [379]:
# calculate linearly interpolated 2035 targets
# base 1990
target2035_base1990 = dict(DE = 0.75,
BE = 0.55 + (1-0.55)*(2035-2030)/(2050-2030),
CH = 0.625,
DK = 0.7 + (1-0.7)*(2035-2030)/(2045-2030),
NL = 0.55 + (1-0.55)*(2035-2030)/(2050-2030),
SE = 0.69,
PL = (esr_targets["PL"]/esr_targets["DE"])*0.75
)
# base 2005
# Polands 2035 target is deduced based on the ESR distrubution key in relation to the German interpolated target
target2035_base2005 = dict(AT = 0.74,
CZ = 0.26 + (1-0.26)*(2035-2030)/(2050-2030),
FR = 0.475 + (1-0.475)*(2035-2030)/(2050-2030),
IT = 0.437 + (1-0.437)*(2035-2030)/(2050-2030),
LU = 0.55 + (1-0.55)*(2035-2030)/(2050-2030)
)
print(target2035_base1990, target2035_base2005)

{'DE': 0.75, 'BE': 0.6625000000000001, 'CH': 0.625, 'DK': 0.7999999999999999, 'NL': 0.6625000000000001, 'SE': 0.69, 'PL': 0.26549999999999996} {'AT': 0.74, 'CZ': 0.445, 'FR': 0.60625, 'IT': 0.57775, 'LU': 0.6625000000000001}


In [380]:
target_2035 = pd.DataFrame(index=co2_totals.index)
target_2035["base1990"] = target2035_base1990
target_2035["base2005"] = target2035_base2005
target_2035

base1990  base2005
Country_code                    
AT                 NaN   0.74000
BE              0.6625       NaN
CH              0.6250       NaN
CZ                 NaN   0.44500
DE              0.7500       NaN
DK              0.8000       NaN
FR                 NaN   0.60625
IT                 NaN   0.57775
LU                 NaN   0.66250
NL              0.6625       NaN
PL              0.2655       NaN
SE              0.6900       NaN

In [381]:
# Calculate aggregated 1990 emissions in MtCO2
co2_1990_aggregated = co2_totals_excl_lulucf.sum()
co2_1990_aggregated

2947.9029606100003

In [382]:
# Calculate absolute target emissions in 2035 based on shares
target_2035["absolute_emissions"] = ((1-target_2035.base1990) * co2_totals_excl_lulucf).combine_first((1-target_2035.base2005) * co2_totals_excl_lulucf_2005)
target_2035_aggregated = target_2035.absolute_emissions.sum()
target_2035_aggregated

1154.8379684469976

In [383]:
# Calculate share of 1990 emissions
round(target_2035_aggregated/co2_1990_aggregated,3)

0.392

Alternate approach: For missing values take interpolated emission reduction target for Germany 2035 (75%). For other countries take distribution key from [EU effort sharing regulation](https://climate.ec.europa.eu/eu-action/effort-sharing-member-states-emission-targets/effort-sharing-2021-2030-targets-and-flexibilities_en). Then calculate total emissions based on 1990 values and weighted emission reduction target.